### PR TITLE
op-node,op-batcher: implement span channel out block count limit

### PR DESCRIPTION
**Pull request by sebastianst**
**Description**

Adds optional support for limiting the number of blocks per span batch. A channel with a span batch block count limit will contain multiple span batches.

**Tests**

Unit tests added, including for the edge case that a multi span batch channel is hitting the exact size limits.

One end-2-end test extended to also test this feature with a max block count of 2.

**Metadata**

- Fixes https://github.com/ethereum-optimism/optimism-private/issues/269

### CODE REVIEW

File: op-e2e/setup.go [hunk 1]
```diff
@@ -288,6 +288,12 @@ type SystemConfig struct {
 	// whether to actually use BatcherMaxL1TxSizeBytes for blobs, insteaf of max blob size
 	BatcherUseMaxTxSizeForBlobs bool
 
+	// Singular (0) or span batches (1)
+	BatcherBatchType uint
+
+	// If 0, limits the number of blocks per span batch
+	BatcherMaxBlocksPerSpanBatch int
```
Comment (axelKingsley) [inline op-e2e/setup.go]:
```suggestion
	// If greater than 0, limits the number of blocks per span batch
	BatcherMaxBlocksPerSpanBatch int
```

File: op-node/rollup/derive/span_channel_out.go [hunk 2]
```diff
@@ -49,7 +57,15 @@ func (co *SpanChannelOut) setRandomID() error {
 	return err
 }
 
-func NewSpanChannelOut(genesisTimestamp uint64, chainID *big.Int, targetOutputSize uint64, compressionAlgo CompressionAlgo, chainSpec *rollup.ChainSpec) (*SpanChannelOut, error) {
+type SpanChannelOutOption func(co *SpanChannelOut)
+
+func SCOWithMaxBlocksPerSpanBatch(maxBlock int) SpanChannelOutOption {
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go]: Personally I am not a fan of such condensed names as `SCO`. Not sure I have a better idea though. Maybe the SCO could be dropped and it'd just be called `WithMaxBlocksPerSpanBatch`, if that doesn't messy up the package namespace.

File: op-node/rollup/derive/span_channel_out.go [hunk 3]
```diff
@@ -166,20 +194,24 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
 	if err = co.compress(); err != nil {
 		return err
 	}
-	co.lastCompressedRLPSize = co.activeRLP().Len()
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go:169]: I see this is moved into the call to `compress`. I think that's probably a better place for it. Trying to remember why i had it up here in the first place.

File: op-node/rollup/derive/span_channel_out.go [hunk 4]
```diff
@@ -166,20 +194,24 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
...
 
 	// if the channel is now full, either return the compressed data, or the compressed previous data
 	if err := co.FullErr(); err != nil {
-		// if there is only one batch in the channel, it *must* be returned
-		if len(co.spanBatch.Batches) == 1 {
+		// if it's the first batch of the channel, it *must* fit in
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go]:
```suggestion
		// if it's the first batch of the channel *and* that batch only contains one block, it *must* fit in
```

technically you could find yourself with only one batch and have multiple blocks inside it and you would would in that case want to roll back the last block of the batch. The code logic is correct, just the comment indicates that batch-count is all that matters.

File: op-node/rollup/derive/span_channel_out.go [hunk 5]
```diff
@@ -166,20 +194,24 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
...
+		if co.sealedRLPBytes == 0 && co.spanBatch.GetBlockCount() == 1 {
+			return nil
+		}
+
+		// if we just perfectly filled up the channel, also return nil to retain block
+		if uint64(co.compressor.Len()) == co.target {
 			return nil
 		}
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go:208]: nice catch :)

For other reviewers - returning nil here is a good idea because there's nothing to roll back. On the *next* attempted addition, or on any fullness check, the Channel Out will report full.

File: op-node/rollup/derive/span_channel_out.go [hunk 6]
```diff
@@ -188,13 +220,36 @@ func (co *SpanChannelOut) AddSingularBatch(batch *SingularBatch, seqNum uint64)
...
 
+func (co *SpanChannelOut) ensureOpenSpanBatch() {
+	if co.maxBlocksPerSpanBatch == 0 || co.spanBatch.GetBlockCount() < co.maxBlocksPerSpanBatch {
+		return
+	}
+	// we assume that the full span batch has been written to the last active rlp buffer
+	active, inactive := co.activeRLP(), co.inactiveRLP()
+	if inactive.Len() > active.Len() {
+		panic("inactive rlp unexpectedly larger")
+	}
```
Comment (axelKingsley) [inline op-node/rollup/derive/span_channel_out.go:231]: Was this something you encountered?

File: op-node/rollup/derive/channel_out_test.go [hunk 7]
```diff
@@ -219,42 +230,46 @@ func TestBlockToBatchValidity(t *testing.T) {
...
+			t.Run(funcName(test)+"_"+algo.String(), func(t *testing.T) {
+				test(t, algo)
 			})
 		}
 	}
 }
 
+func funcName(fn any) string {
+	return runtime.FuncForPC(reflect.ValueOf(fn).Pointer()).Name()
+}
```
Comment (axelKingsley) [inline op-node/rollup/derive/channel_out_test.go:271]: I don't understand the use of `funcName`. Doesn't the run contain the name of the function already just by how `go test` operates?

Are there any issues with `FuncForPC` in other runtimes?

**Review by axelKingsley — APPROVED** [review]
This is a really neat modification to the Span Channel Out system. Where previously Span Channel Out was a totally *non*-streaming structure, now there is a streaming aspect - when a span-batch is full, it can be left in the accumulating buffer that the Channel Out is building. To account for this, the buffers are only reset back to the last full batch. You could think of the previous buffer-reset as being a Truncate(0), because there was only ever one batch and it would never be full.

Once a span batch is full, its bytes are left in the buffer and the object is reset, allowing another batch to stack up on it. This is almost exactly like the singular batch mechanism, but with the iterative building that span batches require.

Left a couple comments, really just about `//comments` or questions. LGTM!

**Review by ajsutton — APPROVED** [review]
I'm leaning on Axel's review as I'm not the most familiar with this code but it LGTM.

File: op-node/rollup/derive/span_channel_out.go (hunk 4 above)
Comment (sebastianst) [reply op-node/rollup/derive/span_channel_out.go]: "batch" here refers to the _singular_ batch, stored in variable `batch`, which corresponds 1:1 to a single block. Will clarify by adding "singular".

File: op-node/rollup/derive/span_channel_out.go (hunk 3 above)
Comment (sebastianst) [reply op-node/rollup/derive/span_channel_out.go:169]: yeah it felt better to move the modification of the state variable into the method that owns the modification. Did the same for the `sealerRLPBytes`, setting it in `ensureOpenSpanBatch`.

File: op-node/rollup/derive/span_channel_out.go (hunk 6 above)
Comment (sebastianst) [reply op-node/rollup/derive/span_channel_out.go:231]: Actually no, I just added this in an early implementation to assert correct code logic, but never hit this panic. We could remove it, given the new unit test coverage.

File: op-node/rollup/derive/span_channel_out.go (hunk 2 above)
Comment (sebastianst) [reply op-node/rollup/derive/span_channel_out.go]: That's fair, I also didn't like it 😅 removed the `SCO`.

File: op-node/rollup/derive/channel_out_test.go (hunk 7 above)
Comment (sebastianst) [reply op-node/rollup/derive/channel_out_test.go:271]: `go test` reads the func name of the top-level test (here `TestSpanChannelOut`), but not of any functions passed to `t.Run`, as these could be anything, also anonymous functions. I just did this to avoid manually writing out function names like it was before, and guarantee correct test names. Before it was
```go
		{"SpanChannelOutCompressionOnlyOneBatch", SpanChannelOutCompressionOnlyOneBatch},
		{"SpanChannelOutCompressionUndo", SpanChannelOutCompressionUndo},
		{"SpanChannelOutClose", SpanChannelOutClose},
```
which IMO is unnecessary if it can be resolved using reflection and runtime information. There are many other places in our codebase where we do this error prone copying, we could slowly transition towards using `funcName`.

File: op-node/rollup/derive/span_channel_out.go (hunk 6 above)
Comment (axelKingsley) [reply op-node/rollup/derive/span_channel_out.go:231]: fine to leave it in, was just curious

File: op-node/rollup/derive/channel_out_test.go (hunk 7 above)
Comment (axelKingsley) [reply op-node/rollup/derive/channel_out_test.go:271]: There's some argument to be made that a `map[string]testFn` is simpler than reflection, but I agree that reflection is the more general (and more cool) solution :D. I'm fine with this

Comment (sebastianst) [reply op-node/rollup/derive/channel_out_test.go:271]: Coolness was admittedly the ultimate reason I did it 😎
